Restore build for old frontends

kunjsong01 opened the pull request:
The commits in this PR restored build for old frontends on top of ESBMC master 0dd4e107734 (10/01/2022). 

However, they stopped working with latest ESBMC master (ea8da5eb2b) due to linker errors: 
```
/usr/bin/ld: src/ansi-c/cpp/libansiccpp.a(hooks.c.o):(.data.rel+0x68): undefined reference to `pthread_buf'
/usr/bin/ld: src/ansi-c/cpp/libansiccpp.a(hooks.c.o):(.data.rel+0x70): undefined reference to `pthread_buf_size'
/usr/bin/ld: src/ansi-c/cpp/libansiccpp.a(hooks.c.o):(.data.rel+0x80): undefined reference to `pthreadtypes_buf'
/usr/bin/ld: src/ansi-c/cpp/libansiccpp.a(hooks.c.o):(.data.rel+0x88): undefined reference to `pthreadtypes_buf_size'
```

My conjecture is that the linker reported undefined reference to 'pthread_*' symbols because it cannot find the generated object files. A further examination of the build folder reveals that the pthread_* files are renamed: 
```
kunjian@kunjian-VirtualBox:~/ESBMC_Project/esbmc/build$ find . -name "pthread.*"
./src/c2goto/headers/pthread.hs.c
./src/c2goto/headers/CMakeFiles/c2gotoheaders.dir/pthread.hs.c.o
./src/c2goto/headers/pthread.hs.h
```

while they used to be: 
```
./src/c2goto/headers/pthread.c
./src/c2goto/headers/CMakeFiles/c2gotoheaders.dir/pthread.c.o
```

@fbrausse Could this be an unexpected side effect of the mangle_hs changes in commit cb442ce6d14a6?

kunjsong01 committed:
Restored old frontend - Part I:

  - replaced the old message system with the new one
  - updated languaget interfaces

kunjsong01 committed:
Restored old frontend - Part II:

  - Use iostream for error reporting in c_main for ansi_c
    for simplicity and quick restoration.

kunjsong01 committed:
Restored old frontend - Part III:

  - added explicit constructors in old ansi-c and cpp frontend

fbrausse commented:
> @fbrausse Could this be an unexpected side effect of the mangle_hs changes in commit cb442ce?

Indeed. I did not touch the old frontends. What CMake flags are you using?

fbrausse commented on src/ansi-c/c_main.cpp:
```diff
@@ -81,8 +81,7 @@ bool c_main(
   {
-    messaget message(message_handler);
-    message.error("main symbol `" + themain + "' not found");
+    std::cout << "main symbol `" << themain << "' not found" << std::endl;
```
Why not use `default_message` from <util/message/default_message.h>?

fbrausse commented on src/ansi-c/c_main.cpp:
```diff
@@ -94,7 +93,7 @@ bool c_main(
     else
     {
-      message.error("main symbol `" + themain + "' is ambiguous");
+      std::cout << "main symbol `" << themain << " is ambiguous" << std::endl;
```
Same: `default_message`?

fbrausse commented on src/cpp/parse.cpp:
```diff
@@ -296,7 +297,7 @@ bool Parser::SyntaxError()
     message += "'";
 
-    parser->print(1, message, location);
+    parser->print((VerbosityLevel)1, message, location);
```
Can we use the named constant `VerbosityLevel::Error` instead?

fbrausse commented on src/cpp/parse.cpp:
```diff
@@ -6118,7 +6119,7 @@ bool Parser::rTryStatement(codet &statement)
         location.set_line(i2string(op.line_no));
 
-        parser->print(1, message, location);
+        parser->print((VerbosityLevel)1, message, location);
```
Same: `VerbosityLevel::Error`

fbrausse commented on src/util/parser.cpp:
```diff
@@ -35,5 +36,5 @@ void parsert::parse_error(const std::string &message, const std::string &before)
   if(before != "")
     tmp += " before `" + before + "'";
-  msg.print(VerbosityLevel::Error, tmp, location);
+  print((VerbosityLevel)1, tmp, location);
```
Why can we not use the named constants?

fbrausse commented on src/util/parser.h:
```diff
@@ -26,12 +26,19 @@ class parsert
 
-  virtual ~parsert() = default;
+  //virtual ~parsert() = default;
+  ~parsert() override = default;
```
Why not remove this defaulted definition (and the commented one)?

fbrausse commented on src/util/parser.h:
```diff
@@ -87,7 +94,7 @@ class parsert
 
 private:
-  const messaget &msg;
+  //const messaget &msg;
```
Can we remove this line?

kunjsong01 commented:
@fbrausse Thanks for the confirmation. 

> What CMake flags are you using?

Here's the list of CMake flags I was using. The old frontend was enabled using '-DENABLE_OLD_FRONTEND=On \'. 
```
cd esbmc && \
mkdir build && \
cd build && \
cmake .. -GNinja \
-DBUILD_TESTING=On \
-DBUILD_STATIC=On \
-DCMAKE_BUILD_TYPE=Debug \
-DENABLE_REGRESSION=On \
-DENABLE_OLD_FRONTEND=On \
-DClang_DIR=$PWD/../../clang11 \
-DLLVM_DIR=$PWD/../../clang11 \
-DBoolector_DIR=$PWD/../../boolector-release \
-DZ3_DIR=$PWD/../../z3 \
-DENABLE_MATHSAT=ON -DMathsat_DIR=$PWD/../../mathsat \
-DENABLE_YICES=On -DYices_DIR=$PWD/../../yices \
-DCVC4_DIR=$PWD/../../cvc4 \
-DGMP_DIR=$PWD/../../gmp -DBitwuzla_DIR=$PWD/../../bitwuzla-release \
-DCMAKE_INSTALL_PREFIX:PATH=$PWD/../../release
```

fbrausse commented:
Many thanks, errors confirmed. This patch fixes the linker errors (by including the generated header), but (as the commit message says) still requires fixing up the filename extension from `.hs` to `.h`: 
[0001-ansi-c-fix-linker-error-file-extension-.hs-still-nee.txt](https://github.com/esbmc/esbmc/files/7880959/0001-ansi-c-fix-linker-error-file-extension-.hs-still-nee.txt)

Edit: why do these headers have ".hs" in the first place?

lucasccordeiro commented on src/ansi-c/c_main.cpp:
```diff
@@ -81,8 +81,7 @@ bool c_main(
   {
-    messaget message(message_handler);
-    message.error("main symbol `" + themain + "' not found");
+    std::cout << "main symbol `" << themain << "' not found" << std::endl;
```
shall we use the fmt library?

kunjsong01 replied:
Fixed and replaced them with the standard message handler. However, using the string concatenation instead of fmt. 

Tried to use fmt headers in the old frontend, but compiler failed to locate them. I've ask @rafaelsamenezes for the most elegant way to include fmt in old frontend. I'll add the fmt::format in my final patch.

kunjsong01 replied:
The fmt library does not link in old frontends. 
Since the string concatenation is a standard C++ operation, I guess it's ok to use the string concatenation instead of fmt.

lucasccordeiro commented on src/ansi-c/convert_character_literal.cpp:
```diff
@@ -36,7 +36,7 @@ void convert_character_literal(const std::string &src, exprt &dest)
 
   typet type;
-  irep_idt cpp_type;
+  //irep_idt cpp_type;
```
shall we remove this line?

lucasccordeiro commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -21,6 +21,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/std_expr.h>
 #include <util/std_types.h>
 #include <util/string_constant.h>
+#include <iostream>
```
we should use the fmt library.

lucasccordeiro commented:
There are some compiling errors in our Linux CI:

```
/home/runner/work/esbmc/esbmc/src/util/context.h:41:39: error: unused parameter ‘t’ [-Werror=unused-parameter]
```

rafaelsamenezes commented:
> Edit: why do these headers have ".hs" in the first place?

@fbrausse  It's for historical reasons. It meant `header script`, as it was used to generate GOTO models

kunjsong01 committed:
Franz's patch: fix old frontend build

fbrausse commented on src/ansi-c/ansi_c_language.cpp:
```diff
@@ -151,13 +155,18 @@ bool ansi_c_languaget::parse(
+#endif
+  // Restore the old_frontend to a working state in Liunx OS.
+  // set parser mode to GCC by default.
+  ansi_c_parser.mode = ansi_c_parsert::GCC;
```
Typo in the comment. You can use config.ansi_c.target.is_windows_abi() to conditionally set the mode to MSC in line with the previous behaviour.

fbrausse commented on src/cpp/cpp_language.cpp:
```diff
@@ -152,13 +156,18 @@ bool cpp_languaget::parse(
+#endif
+  // Restore the old_frontend to a working state in Liunx OS.
+  // set parser mode to GCC by default.
+  cpp_parser.mode = cpp_parsert::GCC;
```
Same comments as for ansi-c/ansi_c_language.cpp

fbrausse commented on src/util/context.h:
```diff
@@ -38,6 +38,11 @@ class contextt
+  contextt& operator=(const contextt& t)
+  {
+    // copy assignment operator for old frontend typechecking
+    return *this;
```
Having a no-op copy-assignment operator is a Bad Idea. Why is it a no-op?

kunjsong01 replied:
Indeed, a no-op copy-assignment opterator is very bad. 

This copy-assignment operator was deliberately made no-op to satisfy the copy operation in `ansi_c_typecheck` function in `src/ansi-c/ansi_c_typecheck.cpp`: 
```
 37 contextt context1(message_handler), context2(message_handler);
 38   ansi_c_parse_treet ansi_c_parse_tree;
 39
 40   context1 = ns.get_context();
```
The copy assignment operation used to work but got broken by the const reference `const messaget &msg;` introduced when centralising the message handler in commit 4dc847814c2. Essentially, this copy-assignment operator restores the semantics of the operation in line 40. (And in my humble opinion, the original usage of default copy-assignment operator was not a good idea either... but it works). 

If we wanted a proper copy-assignment operator of context class for the old frontends, it would be opening a can of worms. Because we would also need to deal with the associated data structures such as symbolst, ordered_symbolst ... etc. It would require a lots of effort to do so, because not all classes in src/util were written based on the rule of three. 

Apologies for the hacky way used in this patch, but I reckon this is perhaps the most efficient way to restore the old C frontend in a timely manner, as well as avoid opening a can of worms.

fbrausse replied:
I see, many thanks for the explanations. Your answer would make an excellent comment for the operator=. ;)
Do you think it is easier not having `contextt` copyable?

kunjsong01 replied:
>Do you think it is easier not having contextt copyable?

Yes, I believe it's currently doing so. When using new C frontend, `contextt` class has a const ref member, but does not have a copy assignment operator or copy constructor. The compiler will report an error like "Use of deleted copy assignment operator" (or something similar) when someone writes a line of code like line 40. 

Additionally, "=delete" has been added to make it more explicit when reading the code, as well as for confomance to the rule of three.

fbrausse replied:
So, what about a comment of explanation on the no-op operator=? E.g. your reply above. No one is going to check github in order to understand why it is a no-op.

kunjsong01 replied:
Done. 
Added a comment to explain the hacky code.

fbrausse commented on src/util/parser.h:
```diff
@@ -26,12 +26,19 @@ class parsert
     clear();
   }
+  */
+  parsert()
```
Since `parsert` now is a `messaget`, shouldn't there be a way to pass an existing `messaget` object for copy-initialization?

kunjsong01 replied:
Added "=delete" for the copy constructor in the latest patch. 
The parser for old frontend is not meant to be used for such operaiton.

fbrausse commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -14,6 +14,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/expr.h>
 #include <util/i2string.h>
 #include <util/parser.h>
+#include <iostream>
```
Why is `<iostream>` now required for _this_ header?

kunjsong01 committed:
Restored old frontend - Part IV:

  - fixed various issues in review comment

fbrausse commented on src/ansi-c/c_main.cpp:
```diff
@@ -81,20 +81,18 @@ bool c_main(
   {
-    messaget message(message_handler);
-    message.error("main symbol `" + themain + "' not found");
+    message_handler.status("main symbol `" + themain + "' not found");
```
Should be `.error()`.

kunjsong01 replied:
Fixed. 
Good catch. Thanks!

kunjsong01 committed:
Fixed remaining issues in comment

kunjsong01 committed:
Fixed lint using clang-format

kunjsong01 committed:
Added comment for no-op copy assignment operator

lucasccordeiro commented on src/cpp/cpp_language.cpp:
```diff
@@ -152,10 +156,10 @@ bool cpp_languaget::parse(
   cpp_parser.filename = path;
   cpp_parser.in = &i_preprocessed;
-  cpp_parser.set_message_handler(&message_handler);
+  //cpp_parser.set_message_handler(&message_handler);
```
Shall we remove this comment?

lucasccordeiro reviewed (approved):
One minor comment only.

kunjsong01 committed:
Fixed mode

lucasccordeiro commented:
@kunjsong01: could you please rebase this branch?

kunjsong01 committed:
Added two test regression test cases to use --old-frontend

kunjsong01 commented:
>could you please rebase this branch?

Rebased done. 
Also added some regression test cases.

kunjsong01 committed:
Fixed regression test case

kunjsong01 committed:
Move test cases to esbmc-old

kunjsong01 committed:
Enabled old frontend build in build_linux workflow

kunjsong01 committed:
Suppressed build errors in ansi-c old frontend

  The following errors are not fixed rather than suppressed:
    - implicit-fallthrough
    - unused local variable

  Also move old frontend test cases to regression/esbmc.

kunjsong01 commented:
Suppressed build errors in legacy code to cope with our build policy -Werror, including:
- unused-function
- unused-const-variable
- char-subscripts

Note that the following errors are not fixed rather than suppressed:
- implicit-fallthrough
- unused local variable

Created esbmc-old regression for old frontend health.

kunjsong01 committed:
Moved test cases to esbmc-old

kunjsong01 commented:
All tests passed. 
Patch is ready to be merged.

lucasccordeiro commented:
Excellent, @kunjsong01.

Many thanks!
